### PR TITLE
feat(kad-dht): routing table

### PR DESCRIPTION
**Pull request by richard-ramos**
(pointing to a branch just to make it easier to review)

### CODE REVIEW

File: tests/kademlia/testroutingtable.nim [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import ../../libp2p/crypto/crypto
+import ../../libp2p/protocols/kademlia/[routingtable, consts, keys]
+
+proc testKey*(x: byte): Key =
+  var buf: array[IdLength, byte]
+  buf[31] = x
+  return Key(kind: KeyType.Undefined, data: buf)
+
+suite "routing table":
+  test "insert single key in correct bucket":
```
Comment (gmelodie) [inline tests/kademlia/testroutingtable.nim]:
```suggestion
  test "inserts single key in correct bucket":
```

File: tests/kademlia/testroutingtable.nim [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    let selfId = testKey(0)
+    var rt = RoutingTable.init(selfId)
+    let other = testKey(0b10000000)
+    rt.insert(other)
+
+    let idx = bucketIndex(selfId, other)
+    check rt.buckets.len > idx
+    check rt.buckets[idx].peers.len == 1
+    check rt.buckets[idx].peers[0].nodeId == other
```
Comment (gmelodie) [inline tests/kademlia/testroutingtable.nim]: maybe `doAssert` in case `peers.len == 0`

File: tests/kademlia/testroutingtable.nim [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+  test "does not insert beyond capacity":
+    let selfId = testKey(0)
+    var rt = RoutingTable.init(selfId)
+    let targetBucket = 6
+    for _ in 0 ..< k + 5:
+      let rng = crypto.newRng()
```
Comment (gmelodie) [inline tests/kademlia/testroutingtable.nim]: Can we reuse the `rng` so you don't call it every time inside the `for`?

File: libp2p/protocols/kademlia/routingtable.nim [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+proc insert*(rtable: var RoutingTable, nodeId: Key) =
+  if nodeId == rtable.selfId:
+    return # No self insertion
+
+  let idx = bucketIndex(rtable.selfId, nodeId)
+  if idx >= maxBuckets:
+    trace "cannot insert node. max buckets have been reached",
+      nodeId, bucketIdx = idx, maxBuckets
+    return
```
Comment (gmelodie) [inline libp2p/protocols/kademlia/routingtable.nim]: I think we should return a `bool` which would be `false` when it doesn't insert

File: libp2p/protocols/kademlia/routingtable.nim [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  RoutingTable* = ref object
+    selfId*: Key
+    buckets*: seq[Bucket]
+
+proc init*(T: typedesc[RoutingTable], selfId: Key): T =
+  return RoutingTable(selfId: selfId, buckets: @[])
+
+proc bucketIndex*(selfId, key: Key): int =
+  let distance = xorDistance(selfId, key)
+  return distance.leadingZeros
```
Comment (gmelodie) [inline libp2p/protocols/kademlia/routingtable.nim]:
```suggestion
proc bucketIndex*(selfId, key: Key): int =
  xorDistance(selfId, key).leadingZeros
```

File: libp2p/protocols/kademlia/keys.nim [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+proc getBytes*(k: Key): seq[byte] =
+  return
+    case k.kind
+    of KeyType.PeerId:
+      k.peerId.getBytes()
+    of KeyType.Raw, KeyType.Undefined:
+      @(k.data)
+
+template `==`*(a, b: Key): bool =
+  a.getBytes() == b.getBytes()
```
Comment (gmelodie) [inline libp2p/protocols/kademlia/keys.nim]: Question: why is this a `template` instead of a regular `proc`?

File: libp2p/protocols/kademlia/keys.nim [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+proc shortLog*(k: Key): string =
+  return
+    case k.kind
+    of KeyType.PeerId:
+      "PeerId:" & $k.peerId
+    of KeyType.Raw, KeyType.Undefined:
+      $k.kind & ":" & toHex(k.data)
```
Comment (gmelodie) [inline libp2p/protocols/kademlia/keys.nim]:
```suggestion
  case k.kind
  of KeyType.PeerId:
    "PeerId:" & $k.peerId
  of KeyType.Raw, KeyType.Undefined:
    $k.kind & ":" & toHex(k.data)
```

File: libp2p/protocols/kademlia/routingtable.nim [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  if idx >= rtable.buckets.len:
+    # expand buckets lazily if needed
+    rtable.buckets.setLen(idx + 1)
+
+  var bucket = rtable.buckets[idx]
+  let keyx = peerIndexInBucket(bucket, nodeId)
+  if keyx.isSome:
+    bucket.peers[keyx.unsafeValue].lastSeen = Moment.now()
+  else:
+    if bucket.peers.len < k:
```
Comment (gmelodie) [inline libp2p/protocols/kademlia/routingtable.nim]:
```suggestion
  elif bucket.peers.len < k:
```

File: libp2p/protocols/kademlia/routingtable.nim [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    if Moment.now() - p.lastSeen > 30.minutes:
+      return true
+  return false
+
+proc randomKeyInBucketRange*(
+    selfId: Key, bucketIndex: int, rng: ref HmacDrbgContext
+): Key =
+  var raw = selfId.getBytes()
+
+  # zero higher bits
```
Comment (gmelodie) [inline libp2p/protocols/kademlia/routingtable.nim]: found the comment confusing at first glance
```suggestion
  # zero out higher bits
```

File: libp2p/protocols/kademlia/routingtable.nim [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+  for i in lsbStart ..< totalBits:
+    let byteIdx = i div 8
+    let bitInByte = 7 - (i mod 8)
+    let lsbByte = (i - lsbStart) div 8
+    let lsbBit = 7 - ((i - lsbStart) mod 8)
+    let randBit = (randomBuf[lsbByte] shr lsbBit) and 1
+    if randBit == 1:
+      raw[byteIdx] = raw[byteIdx] or (1'u8 shl bitInByte)
+    else:
```
Comment (gmelodie) [inline libp2p/protocols/kademlia/routingtable.nim:126]: I feel like it'd be nice to have a `libp2p/bits.nim` file for bit manipulations like these.

**Review by gmelodie — APPROVED** [review]
Only nits

File: libp2p/protocols/kademlia/keys.nim (hunk 6 above)
Comment (richard-ramos) [reply libp2p/protocols/kademlia/keys.nim]: It's so this comparison gets inlined in code instead having to do a proc call

File: libp2p/protocols/kademlia/keys.nim [hunk 11]
```diff
@@ -0,0 +1,48 @@
+import ../../peerid
+import ./consts
+import chronicles
+import stew/byteutils
+
+type
+  KeyType* {.pure.} = enum
+    Undefined
```
Comment (vladopajic) [inline libp2p/protocols/kademlia/keys.nim]: why have `Undefined` as possibility? can it be avoided?

File: libp2p/protocols/kademlia/routingtable.nim [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+  var allNodes: seq[Key] = @[]
+
+  for bucket in rtable.buckets:
+    for p in bucket.peers:
+      allNodes.add(p.nodeId)
+
+  allNodes.sort(
+    proc(a, b: Key): int =
+      cmp(xorDistance(a, targetId), xorDistance(b, targetId))
+  )
```
Comment (vladopajic) [inline libp2p/protocols/kademlia/routingtable.nim:80]: min heap suits this use case perfectly:

```
for each peer
 minHeap add(distance, p.nodeId)
get count elements form minHeap
```

File: tests/kademlia/testroutingtable.nim [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    check isStale(bucket) == false
+
+  test "randomKeyInBucketRange returns id at correct distance":
+    let selfId = testKey(0)
+    let targetBucket = 3
+    var rid = randomKeyInBucketRange(selfId, targetBucket, rng)
+    rid.kind = KeyType.Undefined
+      # Overriding so we don't use sha for comparing xor distances
+    let idx = bucketIndex(selfId, rid)
+    check idx == targetBucket
```
Comment (vladopajic) [inline tests/kademlia/testroutingtable.nim]: check that ids are not the same?

File: tests/kademlia/testroutingtable.nim [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    let ids = @[testKey(1), testKey(2), testKey(3), testKey(4), testKey(5)]
+    for id in ids:
+      discard rt.insert(id)
+
+    let res = rt.findClosest(testKey(1), 3)
+
+    check:
+      res.len == 3
+      res[0] == testKey(1)
+
```
Comment (vladopajic) [inline tests/kademlia/testroutingtable.nim]: okay, but what if `res[1] ==testKey(5)`? better assert whole things

**Review by vladopajic — APPROVED** [review]
would be nice to have more tests. for example
- `xorDistance` is note tested directly, it's tested indirectly by other functions. having direct test cases help to narrow down when issue happens. 
- insert should test all edge cases as they will likely happen
- find closes with edge cases

File: libp2p/protocols/kademlia/keys.nim (hunk 6 above)
Comment (gmelodie) [reply libp2p/protocols/kademlia/keys.nim]: `{.inline.}`?

File: libp2p/protocols/kademlia/keys.nim (hunk 11 above)
Comment (richard-ramos) [reply libp2p/protocols/kademlia/keys.nim]: I guess more than Undefined, it should be called NotHashed, or similar.
